Third step, first try

Vasyl-Pavlenko opened the pull request:
[DEMO LINK](https://vasyl-pavlenko.github.io/react_todo-app-with-api/)

mykhalenych commented on src/App.tsx:
```diff
@@ -1,24 +1,165 @@
+  const [errorMessage, setErrorMessage] = useState('');
+  const [isError, setIsError] = useState(false);
+
+  const isActiveTodos = todos.some(todo => todo.completed);
```
change to useMemo

mykhalenych commented on src/App.tsx:
```diff
@@ -1,24 +1,165 @@
+  const [isError, setIsError] = useState(false);
+
+  const isActiveTodos = todos.some(todo => todo.completed);
+  const countActiveTodos = todos.filter(todo => !todo.completed).length;
```
here as well

mykhalenych reviewed (changes requested):
<img width="1341" alt="Screenshot 2023-03-16 at 17 11 14" src="https://user-images.githubusercontent.com/53228259/225661201-a072314c-62a4-48f6-8d3d-06fc9ebe8466.png">

your loader doen't work correctly, I got an error, and infinity loader

Vasyl-Pavlenko commented:
I added useMemo.

Also i added errorMessage for case offline, but in the same time i couldn't understand how to make my input not disabled?

I tried to add isLoader(false) into my catch. It's doesn't work

loralevitska commented on src/App.tsx:
```diff
@@ -1,24 +1,176 @@
+      setIsTodoLoaded(true);
+      setTodos(todosFromServer);
+    } catch (error) {
+      setIsError(true);
```
I guess it will fix your infinit loaders. Fix loading everywhere
```suggestion
      setIsError(true);
      setIsTodoLoaded(false);
```

loralevitska commented on src/App.tsx:
```diff
@@ -1,24 +1,176 @@
+  };
+
+  const visibleTodos: Todo[] = useMemo(() => {
+    switch (filterBy) {
+      case FilteredBy.ACTIVE:
+        return todos.filter(todo => !todo.completed);
+      case FilteredBy.COMPLETED:
+        return todos.filter(todo => todo.completed);
+      case FilteredBy.ALL:
+        return [...todos];
+      default:
+        throw new Error('Unexpected filter type');
```
```suggestion
    switch (filterBy) {
      case FilteredBy.ACTIVE:
        return todos.filter(todo => !todo.completed);
        
      case FilteredBy.COMPLETED:
        return todos.filter(todo => todo.completed);
        
      case FilteredBy.ALL:
        return [...todos];
        
      default:
        throw new Error('Unexpected filter type');
```

loralevitska commented on src/components/Header/Header.tsx:
```diff
@@ -0,0 +1,74 @@
+  const handleSubmit = (event: React.FormEvent) => {
+    event.preventDefault();
+
+    if (query.length === 0) {
```
```suggestion
    if (!query.length) {
```

loralevitska commented on src/components/TodoItem/TodoItem.tsx:
```diff
@@ -0,0 +1,134 @@
+
+  return (
+    <div
+      className={classNames('todo', { completed: completed === true })}
```
```suggestion
      className={classNames('todo', { completed: !!completed })}
```

OR
If completed always boolean
```suggestion
      className={classNames('todo', { completed })}
```

loralevitska commented on src/utils/fetchClient.ts:
```diff
@@ -0,0 +1,46 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+const BASE_URL = 'https://mate.academy/students-api';
+
+// returns a promise resolved after a given delay
```
Remove comments everywhere. Try it write obvious code that explain all we need without comments)

loralevitska reviewed (changes requested):
Keep up the good work! Some comments to fix:

- After error disable loading 
![image](https://user-images.githubusercontent.com/70009205/225708655-7d022210-ab6f-4d8b-8fdd-f64298f03707.png)
- Remove default outline 
![image](https://user-images.githubusercontent.com/70009205/225708806-0412da8b-caa5-4f9d-b22e-add0efe06d91.png)
- Your toggleAll doesn't work correct 
- Every adding new todo fetch 2 queris. Should be 1 query per todo. Fix it

If you have any questions, please ask them in fe_chat and tag mentors. It will be faster, because your answers in GitHub will be visible to mentor only after you re-request review

DarkMistyRoom commented on src/App.tsx:
```diff
@@ -1,24 +1,192 @@
+    });
+    try {
+      setIsTodoLoaded(true);
+      await createTodo(USER_ID, data);
+      await getTodosFromServer();
```
would be better for performance to reduce requests on the server and update the array on the place. Same for all other manipulations with todos

DarkMistyRoom commented on src/App.tsx:
```diff
@@ -1,24 +1,192 @@
+
+    try {
+      setIsTodoLoaded(true);
+      await updateTodo(todoId, todoCopy);
+      getTodosFromServer();
```
```suggestion
      await updateTodo(todoId, todoCopy);
      await getTodosFromServer();
```

DarkMistyRoom commented on src/App.tsx:
```diff
@@ -1,24 +1,192 @@
+    });
+  };
+
+  const clearCompletedTodo = async () => {
+    const completedTodos = todos.filter(todo => todo.completed);
+
+    completedTodos.forEach(todo => removeTodoOnServer(todo.id));
+  };
```
what about Loader? you should show it while the request is handling

DarkMistyRoom commented on src/types/SendTodo.ts:
```diff
@@ -0,0 +1,5 @@
+export interface SendTodo {
+  userId: number,
+  title: string,
+  completed: boolean,
+}
```
you can place it in the Todo.ts

DarkMistyRoom reviewed (changes requested):
- I've just added todo with only whitespaces

![image](https://user-images.githubusercontent.com/98764641/225866697-8fe7a872-38ad-46cc-9149-4314522077d0.png)

- Also would be great to add focusing on input on editing

![image](https://user-images.githubusercontent.com/98764641/225872348-d8702c9e-1f75-4d73-9be6-fd509af620c8.png)

- Check how toggle all works in the example page

IvanFesenko commented on src/App.tsx:
```diff
@@ -1,24 +1,219 @@
+
+    try {
+      setIsTodoLoaded(true);
+      await updateTodo(todoId, todoCopy);
+      await getTodosFromServer();
```
This case will gives multiple request. Just update local state without request all todos

IvanFesenko commented on src/App.tsx:
```diff
@@ -1,24 +1,219 @@
+        return todos.filter(todo => todo.completed);
+
+      case FilteredBy.ALL:
+        return [...todos];
```
No need to create copy
```suggestion
        return todos;
```
